concordium-smart-contract-testing/src/lib.rs

```diff
@@ -0,0 +1,2745 @@
+                                                code: reason,
+                                                data: return_value,
+                                            },
+                                        },
+                                        remaining_energy,
+                                        false,
+                                    ),
+                                    v1::ReceiveResult::Trap {
+                                        remaining_energy, ..
+                                    } => (
+                                        false,
+                                        InvokeResponse::Failure {
+                                            kind: v1::InvokeFailure::RuntimeError,
+                                        },
+                                        remaining_energy,
+                                        false,
+                                    ),
+                                    v1::ReceiveResult::OutOfEnergy => (
+                                        false,
+                                        InvokeResponse::Failure {
+                                            kind: v1::InvokeFailure::RuntimeError,
+                                        },
+                                        0,
+                                        false,
+                                    ), // TODO: What is the correct error here?
+                                },
+                                Err(_e) => (
+                                    false,
+                                    InvokeResponse::Failure {
+                                        kind: v1::InvokeFailure::RuntimeError,
+                                    },
+                                    0,
+                                    false,
+                                ), // TODO: Correct energy here?
+                            };
+
+                            // Add resume event
+                            let resume_event = ChainEvent::Resumed {
+                                address: self.address,
+                                success,
+                            };
+
+                            println!(
+                                "\tResuming contract {}\n\t\tafter {}",
+                                self.address,
+                                if success {
+                                    "succesful invocation"
+                                } else {
+                                    "failed invocation"
+                                }
+                            );
+                            self.chain_events.push(resume_event);
+
+                            // Update the mutable state, since it might have been changed on
+                            // reentry.
+                            self.mutable_state =
+                                self.chain.get_instance(self.address)?.state.thaw();
```

Comment:
I think we need to do this using the native checkpointing that the mutable state already has.